### PR TITLE
refactor(handle_click): remove vim script wrapper

### PR DESCRIPTION
**Pull request by nullchilly**
For some reason I failed to make them work with `v:lua.require'bufferline'.` Other than that, works for me :) 
Oops all checks have failed lol. Do I need to rewrite the tests too

Fixes: https://github.com/akinsho/bufferline.nvim/issues/456

### CODE REVIEW

File: lua/bufferline/ui.lua [hunk 1]
```diff
@@ -145,7 +145,7 @@ function M.make_clickable(func_name, id, component)
   -- v:lua does not support function references in vimscript so
   -- the only way to implement this is using autoload vimscript functions
   component.attr = component.attr or {}
-  component.attr.prefix = "%" .. id .. "@nvim_bufferline#" .. func_name .. "@"
+  component.attr.prefix = "%" .. id .. "@v:lua.nvim_bufferline_" .. func_name .. "@"
```
Comment (akinsho) [inline lua/bufferline/ui.lua:148]: There are probably more places than just here probably where this needs to also be changed, please have a look to make sure you got them all

File: lua/bufferline/commands.lua [hunk 2]
```diff
@@ -109,7 +109,7 @@ local cmds = {
 ---Handler for each type of mouse click
 ---@param id number
 ---@param button string
-function M.handle_click(id, button)
+function _G.nvim_bufferline_handle_click(id, _, button)
```
Comment (akinsho) [inline lua/bufferline/commands.lua:112]: This name is probably quite likely to cause a namespace clash. I suggest something much more secure if we are going to pollute the global namespace e.g. create a global table and then put a handle click function inside it e.g.

```lua
_G.___bufferline_private = _G.__bufferline_private or {} -- to guard against reloads
_G.___bufferline_private.handle_click = ...

```

**Review by akinsho — CHANGES_REQUESTED** [review]
@nullchilly thanks for the PR left some comments about improving the naming to make the global variables less easy to accidentally override or discover. 

Also, you will need to update the tests probably you just need to change the names in the tests also

File: lua/bufferline/ui.lua (hunk 1 above)
Comment (nullchilly) [reply lua/bufferline/ui.lua:148]: I searched them through telescope through the keyword `nvim_bufferline#` as that's the only way to call the vimscript I think?

File: lua/bufferline/commands.lua (hunk 2 above)
Comment (nullchilly) [reply lua/bufferline/commands.lua:112]: Where do I put these global variables, at the top?

Comment (akinsho) [reply lua/bufferline/commands.lua:112]: @nullchilly you can put it near the top bufferline.lua file after the imports and maybe add a comment about why it's there

File: lua/bufferline.lua [hunk 3]
```diff
@@ -22,6 +22,10 @@ local constants = lazy.require("bufferline.constants")
 --- @module "bufferline.highlights"
 local highlights = lazy.require("bufferline.highlights")
 
+-- @v:lua@ in tabline only support global function so this is
+-- the only way without autoload vimscript functions
```
Comment (akinsho) [inline lua/bufferline.lua:26]:
```suggestion
-- @v:lua@ in the tabline only supports global functions, so this is
-- the only way to add click handlers without autoloaded vimscript functions
```

File: lua/bufferline.lua [hunk 4]
```diff
@@ -22,6 +22,10 @@ local constants = lazy.require("bufferline.constants")
...
+_G.___bufferline_private = _G.__bufferline_private or {} -- to guard against reloads
```
Comment (akinsho) [inline lua/bufferline.lua:27]:
```suggestion
_G.___bufferline_private = _G.___bufferline_private or {} -- to guard against reloads
```

**Review by akinsho — APPROVED** [review]
LGTM 👍🏿 thanks 🚀

**Comment by nullchilly** [comment]
Since this is unmerged is it possible to reopen the pupl request?

**Comment by akinsho** [comment]
@nullchilly I added the merge commit to the version 3 branch which i will merge when 0,8 is stable

**Pull request merged**